### PR TITLE
Throttle download requests to prevent hitting Github's abuse rate limits

### PR DESCRIPTION
**Pull request by brahma-dev**

### CODE REVIEW

**Comment by jlkalberer** [comment]
This is funny. I was actually thinking about how Github would throttle these requests last night.

---

I think the logic here is hard to follow. A better approach looks like:

1. Filter out all assets we don't care about
2. Convert remaining assets into chunks: `Array<Array<Asset>>`
3. Chain chunk fetching with promises. So we don't download chunk 2 until chunk 1 is finished

This isn't tested code but it would look something like this:
```
const CHUNK_SIZE = 10;

function chunkArray(input, chunkSize) {
	let index = 0;
    const arrayLength = input.length;
    const tempArray = [];
    
    for (index = 0; index < arrayLength; index += chunkSize) {
        tempArray.push(input.slice(index, index+chunkSize));
    }

    return tempArray;
}

const chunks = chunkArray(
	assets.filter(
		(asset: Object): Promise<string> => 
			asset.name.match(/(system-part|bootloader)/)
	),
	CHUNK_SIZE,
);

return chunks.reduce((promise, chunk) => 
	promise.then(results => 
		Promise.all([
			...results,
			...chunk.map(asset => downloadAssetFile(asset))
		])
	)
, Promise.resolve());
```

**Comment by brahma-dev** [comment]
In fact, I had initially thought of doing it in chunks, but went in favour of having 10 files downloading at all times.

**Comment by jlkalberer** [comment]
Ahh got it. So you made 10 "streams" that downloaded in parallel. That's actually really cool but the code was hard to understand :)
